src/main/java/com/google/api/generator/gapic/composer/grpc/GrpcServiceStubClassComposer.java

```diff
@@ -245,10 +255,171 @@ protected String getProtoRpcFullMethodName(Service protoService, Method protoMet
+    TypeNode returnType =
+        TypeNode.withReference(
+            ConcreteReference.builder()
+                .setClazz(Map.class)
+                .setGenerics(TypeNode.STRING.reference(), TypeNode.STRING.reference())
+                .build());
+    Expr returnExpr;
+    // If the google.api.routing annotation is present(even with empty routing parameters),
+    // the implicit routing headers specified in the google.api.http annotation should not be sent
+    if (method.routingHeaders() == null) {
+      returnExpr =
+          addRequestParamsForHttpBindings(method, bodyStatements, requestVarExpr, returnType);
+    } else {
+      returnExpr =
+          addRequestParamsForRoutingHeaders(
+              method, classStatements, bodyStatements, requestVarExpr, returnType);
+    }
 
+    // Overrides extract().
+    // https://github.com/googleapis/gax-java/blob/8d45d186e36ae97b789a6f89d80ae5213a773b65/gax/src/main/java/com/google/api/gax/rpc/RequestParamsExtractor.java#L55
+    return LambdaExpr.builder()
+        .setArguments(requestVarExpr.toBuilder().setIsDecl(true).build())
+        .setBody(bodyStatements)
+        .setReturnExpr(returnExpr)
+        .build();
+  }
+
+  private Expr addRequestParamsForRoutingHeaders(
+      Method method,
+      List<Statement> classStatements,
+      List<Statement> bodyStatements,
+      VariableExpr requestVarExpr,
+      TypeNode returnType) {
+    TypeNode routingHeadersBuilderType =
+        TypeNode.withReference(
+            ConcreteReference.builder().setClazz(RoutingHeaderParamsBuilder.class).build());
+    VariableExpr routingHeadersBuilderVarExpr =
+        VariableExpr.builder()
+            .setVariable(
+                Variable.builder().setName("builder").setType(routingHeadersBuilderType).build())
+            .setIsDecl(true)
+            .build();
+    Expr newBuilderExpr = NewObjectExpr.builder().setType(routingHeadersBuilderType).build();
+    Expr newRoutingHeadersAssignExpr =
+        AssignmentExpr.builder()
+            .setVariableExpr(routingHeadersBuilderVarExpr)
+            .setValueExpr(newBuilderExpr)
+            .build();
+    bodyStatements.add(ExprStatement.withExpr(newRoutingHeadersAssignExpr));
+    List<RoutingHeader> routingHeaders = method.routingHeaders().routingHeadersList();
+    VariableExpr routingHeadersBuilderVarNonDeclExpr =
+        VariableExpr.builder()
+            .setVariable(
+                Variable.builder().setName("builder").setType(routingHeadersBuilderType).build())
+            .build();
+    for (int i = 0; i < routingHeaders.size(); i++) {
+      RoutingHeader routingHeader = routingHeaders.get(i);
+      MethodInvocationExpr requestFieldGetterExpr =
+          createRequestFieldGetterExpr(requestVarExpr, routingHeader.fieldName());
+      Expr routingHeaderKeyExpr =
+          ValueExpr.withValue(StringObjectValue.withValue(routingHeader.name()));
+      // TODO: Create proper snake style upper case name
```

Comment:
You can use `JavaStyle.toUpperSnakeCase()`?

Reply:
Yep, thanks for the suggestion, it work perfectly!